src/http/modules/ngx_http_ssl_module.c

```diff
@@ -788,9 +788,18 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
     if (conf->verify) {
 
         if (conf->client_certificate.len == 0 && conf->verify != 3) {
-            ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                          "no ssl_client_certificate for ssl_verify_client");
-            return NGX_CONF_ERROR;
+            if (conf->protocols & (NGX_SSL_SSLv3|NGX_SSL_TLSv1))
+            {
+                ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
+                            "no ssl_client_certificate for ssl_verify_client");
+                return NGX_CONF_ERROR;
+            }
+            if (conf->trusted_certificate.len == 0) {
+                ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
+                            "no ssl_client_certificate or "
+                            "ssl_trusted_certificatefor ssl_verify_client");
+                return NGX_CONF_ERROR;
+            }
```

Comment:
I don't think it makes sense to separate logic for older protocol versions, because this complicates code for no apparent reason:
- legacy setups will barely change using legacy configs, it should be safe to expand to all protocol versions, even if it might not work
- empty DN list works in practice with older protocol versions, SSL/TLS library handles that for us.
- if any legacy clients exist that cannot accept TLSv1.1 AND empty DN lists, then see the 1st point about legacy setups. 
As such, we can fold a single config check.
```suggestion
        if (conf->verify != 3
            && conf->client_certificate.len == 0
            && conf->trusted_certificate.len == 0)
        {
            ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
                          "no ssl_client_certificate or "
                          "ssl_trusted_certificate for ssl_verify_client");
            return NGX_CONF_ERROR;
        }
```

Reply:
@pluknet Thanks a lot for finding time to review it. I express same concern in [this thread.](https://mailman.nginx.org/pipermail/nginx-devel/2024-August/6VF2IN3A24UZKVLQ3IUUL7AJY6L5QS2Z.html). 

Where I asked below. So, I already agree with you.
```
Also, let me know if we need to keep
check simple
as this:
*if (conf->client_certificate.len == 0 && conf->trusted_certificate.len ==
0 && conf->verify != 3) {*
```

Reply:
@pluknet I will reraise-PR with suggested commit today.  

AFAIK, as per contributing guide, in this case, it has to be a forced push to make it one commit. Kindly let me know, if we prefer "Squash and Merge" instead, which can have 2 commits, i.e. 2nd commit to address review comments.

Reply:
Generally we ask to break changes in logically-separate self-sufficient commits to maintain a clean history.
As per CONTRIBUTING.md:
```
- Keep a clean, concise and meaningful commit history on your branch, rebasing
locally and breaking changes logically into commits before submitting a PR
```
So, you were correct to make a force push with a single commit.